File: adaptive_repair/data/python_programs/next_permutation.py
def next_permutation(perm):
    for i in range(len(perm) - 2, -1, -1):
        if perm[i] < perm[i + 1]:
            # Found the pivot 'i'. Now find the smallest element in the suffix perm[i+1:]
            # that is greater than perm[i]. Since perm[i+1:] is in descending order,
            # iterating from right to left and finding the first element perm[j] such
            # that perm[j] > perm[i] will correctly identify this element.
            for j in range(len(perm) - 1, i, -1):
                if perm[j] > perm[i]:  # BUG FIX: Changed condition from < to >
                    next_perm = list(perm)
                    next_perm[i], next_perm[j] = next_perm[j], next_perm[i]
                    # After swapping, the suffix perm[i+1:] needs to be sorted in ascending order.
                    # Since it was in descending order before the swap, reversing it achieves this.
                    next_perm[i + 1:] = list(reversed(next_perm[i + 1:])) # Ensure slice assignment uses a list
                    return next_perm
    # According to the precondition, this part should not be reached.
    # If it were, it would mean the input perm is sorted in reverse order,
    # and the next permutation would be the sorted (ascending) list.
    return None # Or raise an error, or return sorted(perm) if precondition was relaxed.


"""
Next Permutation
next-perm


Input:
    perm: A list of unique ints

Precondition:
    perm is not sorted in reverse order

Output:
    The lexicographically next permutation of the elements of perm

Example:
    >>> next_permutation([3, 2, 4, 1])
    [3, 4, 1, 2]
"""

File: adaptive_repair/data/java_programs/NEXT_PERMUTATION.java
package java_programs;

public class NEXT_PERMUTATION {
    public static int[] next_permutation(int[] perm) {
        int n = perm.length;
        int i = n - 2;
        while (i >= 0 && perm[i] >= perm[i + 1]) {
            i--;
        }

        if (i == -1) {
            return null; // Equivalent to Python's [] to signal no next permutation
        }

        int j = n - 1;
        // Find the largest index j to the right of i such that perm[j] > perm[i].
        // Due to the descending order of perm[i+1:], this also finds the smallest value perm[j] > perm[i].
        while (j > i) {
            if (perm[j] > perm[i]) {
                break;
            }
            j--;
        }

        // Swap perm[i] and perm[j]
        int temp = perm[i];
        perm[i] = perm[j];
        perm[j] = temp;

        // Reverse the suffix in-place using two pointers to avoid creating new list objects
        int left = i + 1;
        int right = n - 1;
        while (left < right) {
            // Swap perm[left] and perm[right]
            temp = perm[left];
            perm[left] = perm[right];
            perm[right] = temp;
            left++;
            right--;
        }

        return perm;
    }
}